From 3d8469d200c8ff235ee59150c437ed87959a7c2c Mon Sep 17 00:00:00 2001
From: Corentin Oparowski <corentin.oparowski@vates.tech>
Date: Fri, 9 May 2014 11:40:46 +0100
Subject: [PATCH 10/35] x86,xen: correct dma_get_required_mask() for Xen PV
 guests

On systems where DMA addresses and physical addresses are not 1:1
(such as Xen PV guests), the generic dma_get_required_mask() will not
return the correct mask (since the xen swiotlb does not provide a
get_required_mask callback, it falls back to using 32 bits).

Some device drivers (such as mpt3sas) use dma_get_required_mask() to set
device DMA masks to allow them to use only 32-bit DMA addresses in
hardware structures.  This results in unnecessary use of the SWIOTLB if
DMA addresses are more than 32-bits, impacting performance
significantly.

Under Xen, the required DMA mask can then always be set to 64-bits.

Signed-off-by: David Vrabel <david.vrabel@citrix.com>
---
 drivers/xen/swiotlb-xen.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/xen/swiotlb-xen.c b/drivers/xen/swiotlb-xen.c
index ef56a2500ed6..534faa440f71 100644
--- a/drivers/xen/swiotlb-xen.c
+++ b/drivers/xen/swiotlb-xen.c
@@ -404,6 +404,12 @@ xen_swiotlb_dma_supported(struct device *hwdev, u64 mask)
 	return xen_phys_to_dma(hwdev, default_swiotlb_limit()) <= mask;
 }
 
+static u64
+xen_swiotlb_get_required_mask(struct device *dev)
+{
+	return DMA_BIT_MASK(64);
+}
+
 const struct dma_map_ops xen_swiotlb_dma_ops = {
 #ifdef CONFIG_X86
 	.alloc = xen_swiotlb_alloc_coherent,
@@ -425,5 +431,6 @@ const struct dma_map_ops xen_swiotlb_dma_ops = {
 	.get_sgtable = dma_common_get_sgtable,
 	.alloc_pages_op = dma_common_alloc_pages,
 	.free_pages = dma_common_free_pages,
+	.get_required_mask = xen_swiotlb_get_required_mask,
 	.max_mapping_size = swiotlb_max_mapping_size,
 };
-- 
2.47.3

